evennia/contrib/turnbattle.py

```diff
@@ -0,0 +1,520 @@
+
+BASE_CHARACTER_TYPECLASS = "evennia.contrib.turnbattle.BattleCharacter"
+CMDSET_CHARACTER = "evennia.contrib.turnbattle.CharacterCmdSet"
+
+If you want to expand upon this system, it's recommended you copy the
+code over to your game's modules and import from there instead.
+"""
+
+from evennia import Command
+from evennia import default_cmds
+from typeclasses.scripts import Script
+from typeclasses.characters import Character
+from commands.command import Command
+from random import randint
+
+class BattleCharacter(Character):
+    
+    def at_object_creation(self):
+        """
+        Adds attributes for a character's current and maximum HP.
+        We're just going to set this value at '100' by default.
+        
+        You may want to expand this to include various 'stats' that
+        can be changed at creation and factor into combat calculations.
+        """
+        self.db.max_hp = 100
+        self.db.hp = self.db.max_hp
+    pass
+    def at_before_move(self, destination):
+        """
+        This keeps characters from moving when in combat or at 0 HP.
+        """
+        if is_in_combat(self):
+            self.caller.msg("You can't exit a room while in combat!")
+            return False # Returning false keeps the character from moving.
+        if self.db.HP <= 0:
+            self.caller.msg("You can't move, you've been defeated!")
+            return False
+        return True
+    
+class CharacterCmdSet(default_cmds.CharacterCmdSet):
+    """
+    Adds combat commands to the default command set.
+    """
+    key = "DefaultCharacter"
+
+    def at_cmdset_creation(self):
+        """
+        Populates the cmdset
+        """
+        super(CharacterCmdSet, self).at_cmdset_creation()
+        self.add(CmdFight())
+        self.add(CmdAttack())
+        self.add(CmdRest())
+        self.add(CmdPass())
+        self.add(CmdDisengage())
+        
+"""
+----------------------------------------------------------------------------
+COMMANDS START HERE
+----------------------------------------------------------------------------
+"""
+class CmdFight(Command):
+    """
+    Starts a fight with everyone in the same room as you.
+    
+    Usage:
+      fight
+    
+    When you start a fight, everyone in the room who is able to
+    fight is added to combat, and a turn order is randomly rolled.
+    When it's your turn, you can attack other characters.
+    """
+    key = "fight"
+    help_category = "combat"
+
+    def func(self):
+        """
+        This performs the actual command.
+        """
+        here = self.caller.location
+        fighters = []
+        
+        if not self.caller.db.hp: # If you don't have any hp
+            self.caller.msg("You can't start a fight if you've been defeated!")
+            return
+        for thing in here.contents: # Test everything in the room to add it to the fight.
+            if thing.db.HP: # If the object has HP...
```

Comment:
This is a nice touch, it means you could fight a table or door assuming it has HP. :)